add LocationIndexTree.query(BBox)

karussell opened the pull request:
This PR adds a new query(BBox) method. Currently we can fetch only nodes from one leaf and so with this querying areas is inefficient. Related to #1127 and #1324.

Furthermore a possibility to visualize the index 

![image](https://user-images.githubusercontent.com/129644/47011326-f9461800-d141-11e8-8f89-da3dd2419730.png)

and some description for the index is added. The description now makes clear that we have a Quadtree implementation with some special tuning. E.g. for bigger areas one node branches not only into 4 subtrees but can branch into 16. Previously also 64 was possible, but we removed this as this has no advantages in terms of storage, query or creation speed. (Measured it for Germany and world wide)

TODOs:

 - [x] ~~to make this correct we need to add +1/2 cell on every side of the BBox. The problem is that we do not add the node to both cells due to the [bresenham line to cell limitation](http://stackoverflow.com/q/24679963/194609)~~ Does not seem to be necessary (no problems e.g. in #1572).
 - [ ] make it working with Polygon too (use JTS Polygon e.g. via embedding in ours so that we can implement `Shape`)
 - [x] use PrepareGeometry  -> #1589
 - [ ] use intersect and contains!
 - [ ] use quality vs. speed properties like UNIQUE_NODES (just make sure nodes come to the Visitor just once), FILTERED_NODES (unique nodes plus filtered again through the shape), HIGH_PRECISION (filtered node but use fetchGeometry instead of just junction nodes)

karussell commented on reader-osm/src/test/java/com/graphhopper/reader/osm/GraphHopperOSMTest.java:
```diff
@@ -149,6 +145,67 @@ public void testLoadOSMNoCH() {
+                if (bbox.contains(lat, lon))
+                    indexNodeList.add(nodeId);
+
+                // TODO should be done behind the scene?
```
Will make it possible via a `SimpleVisitor` class or something. Often this iteration has to be done again on the client side e.g. to fetch the edges and so this work would be an unnecessary overhead then.

michaz commented:
Nice, this closes some open ends for me. Is the current state mergeable, meaning, do we think that the bounding-box query does more or less what it should? Because I'd like to.

karussell commented:
Well, this is not fully ready unfortunately. Currently all nodes are returned where the tile intersects with the shape. This is ok for visualisation, but for other use cases we have to filter these nodes and call "shape.contains(node)" as a post processing step. Now the problem is that this could be done cheap like this or in a more expensive manner if we need high precision via "shape.intersects(edge.fetchWayGeometry)". Currently exploring the options on how to do this so that this also works for a Polygon (which basically means using JTS Polygon behind our Polygon class). But we could also bring it into a mergable state faster and I'll create a new issue for the open tasks.

michaz commented:
Oh, that's exactly what I need: "May include more nodes than are contained, but never fails to return something that _is_ contained". I think that's also how other implementations work.

karussell commented:
> Oh, that's exactly what I need: "May include more nodes than are contained, but never fails to return return a node that is contained". I think that's also how other implementations work.

Ah, ok. I think there are indeed two use cases. This one, which you need and we need for fast visualization. But also the other to e.g. block certain edges but not too many!

Maybe I'll bring this PR in a mergable form for the first use case and explicitly mention this and work later on the second "high precision" use case?

michaz commented:
> Maybe I'll bring this PR in a mergable form for the first use case and explicitly mention this and work later on the second "high precision" use case?

Yes, please! For me, this doesn't even mean that this is currently "unfinished" -- _filtering_ the result to the desired level of exclusivity is a _second_ task, and can happen outside of the current method. This one is finished.

michaz commented:
And this method can still narrow down what it returns later on without changing its promise.

karussell commented:
> filtering the result to the desired level of exclusivity is a second task

This is exactly where I'm unsure because the method `index.query(shape)` could also imply that this should be an exact filter and no post-processing work needs to be done.

michaz commented:
Well, "query" as a method name is vague enough that I think it can't be misunderstood as promising anything specific. ;-)

Let's add a Javadoc which explains what this method does and does not (not _how_ it does that), and it'll be fine.

karussell commented:
Ok :D 

Will do this ... let me just do some minor stuff.

karussell commented:
I did the following changes (let me know if this was too much):

 * moved the Visitor interfaces to the LocationIndex interface
 * moved the query method to the interface
 * removed the extra visualize method and made it possible via the Visitor setting isTileInfo (now Visitor has to be an abtract class as default methods are not possible in 7). onTile now directly includes the depth not the too-visualization-specific `width`.
 * avoid duplicate nodes (and edges for the EdgeVisitor)
 * introduced a new "queryBBox.contains" branch to avoid bbox checks on greater depths, should make the query slightly faster
 * renamed Shape.intersect to intersects, also to make later migrations to JTS easier
 * renamed `*cell*` to `*tile*`
 * made some methods non-public

michaz commented on core/src/main/java/com/graphhopper/storage/index/LocationIndex.java:
```diff
@@ -62,4 +67,56 @@
+    /**
+     * This interface allows to visit every node stored in the leafs of a LocationIndex.
+     */
+    abstract class Visitor {
```
Just as a small remark: This Visitor is not very Java-8-friendly, and also a bit unusual. If I understand it correctly, as a client, I have to override `isTileInfo` if I want `onTile` to be called? I think interfaces are most clear when control flows in one direction only: An interface is either for me to call, or for me to implement and be called.

michaz replied:
> An interface is either for me to call, or for me to implement and be called.

Okay, that's actually the case here. :-) But the _information_ flows both ways.

karussell replied:
Yes, I tried with a separate QueryOptions class but this was not much prettier. Should I revert it back to this to make it easier with lambda?

michaz replied:
Ah no, it's okay.

michaz replied:
Thanks! I'm doing more Isochrone experiments with this right now... ,-)
